Map editor

Nemikolh opened the pull request:
## Map editor
- [x] webgl component
- [x] Renderer
  - [x] a huge number of tiles
  - [x] a sprite of any size and only one texture
  - [x] basic API to draw stuff with webgl
  - [x] possibility to apply camera based overlay effects
  - [ ] ~~support object rendering~~
  - [ ] ~~heightmap sort~~
- [ ] ~~options~~
  - [x] show grid
  - [ ] ~~show heightmap~~
- [ ] ~~add a matrix library (or quick craft one for our needs)~~
- [ ] ~~back-end support for chipset~~
- [ ] ~~dialog to upload chipset~~
  - [ ] ~~possibility to split into tiles~~
  - [ ] ~~possibility to select an area and mark it as an object~~
  - [ ] ~~for simple object: define the blocking area~~
- [ ] ~~commit a map~~
- [ ] ~~save a map~~
- [x] create a map
- [ ] ~~collaboration on the map~~
- [x] zoom
- [x] pan
- [ ] ~~create/order layers~~
- [x] edition
  - [x] replace tile of a layer by ones contains by the brush
  - [ ] ~~support brush size greater than one tile~~
  - [x] change the content of the brush
- [x] palette
  - [x] open the palette
  - [x] edit brush
  - [ ] ~~text-search for virtual chipset~~

Nemikolh committed:
Added MapEditor to toolbar. Basic css to position the canvas.

Nemikolh committed:
[WIP] Schemas for maps.

Nemikolh committed:
Merge remote-tracking branch 'origin/master' into map-ed

Conflicts:
	editor/components/map/editor.ts
	editor/components/toolbar/toolbar.html

Nemikolh committed:
Upgraded few libs, added work in progress webgl code for the map.

Nemikolh committed:
Finished the webgl component.

Nemikolh committed:
Added support for textures.

Nemikolh commented on server/apis/auth.ts:
```diff
@@ -1,20 +1,38 @@
+            }
+        });
+    })(req, res, next);
+}, logRequest('has logged in.'));
```
Erf, I should probably have put that in another PR. :/

Nemikolh committed:
Switched to NEAREST filtering.

Nemikolh committed:
Texture contains image size. Prepared the transition towards RenderingContext abstraction.

Nemikolh committed:
Explicit TypeScript version we are currently using.

Nemikolh committed:
Fixed blending and moved program out of draw call to call it once for multiple objects.

Nemikolh committed:
Camera zoom is kinda working. The API needs to be changed and to accomodate for tiles.

Nemikolh committed:
Add a TilesLayer to manage layers of same size.

This is work in progress but add the basic functionality
to render map or big object that are made of tiles.

The TilesLayer handles:

 * multiple layers with same size and same tile size.
 * a unique vertex buffer and indices buffer for each layer.
 * the update of the buffers when the camera change.

Each Layer is made of PartialLayer that handle:

 * a texture (chipset)
 * a list of tile ids (zero reserved for discarded tiles)
 * the final rendering

Nemikolh committed:
Finish the TilesLayer 'draw()' impl

This commit has a correct and final version
that update when the camera has changed and
performs the correct transformation assuming
the data is correct.

 * We use and index buffer but this is debatable
   as it only starts to be memory efficient when
   using more than one texture or more than one
   layer.

 * We could also have an even less memory saving
   view where each `PartialLayer` has its own
   `VertexBuffer`. This would avoid the -1 in the
   vertex shader and all the degenerates cases
   generated by vertices that do not have a valid
   texture id.

Nemikolh committed:
Improve the VertexBuffer interface regarding data type.

Nemikolh committed:
Add builder API for TilesLayer

This commit also includes an updated version of the
documentation that explains the reasoning.
Using more interfaces, we make use of the type system
to enforce ordering in builder calls.

Nemikolh committed:
Adapt TilesRenderingContext for Textures.

Nemikolh committed:
Add working version of the TilesReneringContext

* It turns out that `uint` and `gl_VertexID` can't be used in GLSL ES.
  So we have to generate the texture coordinates in TypeScript.
  This essentially means that we will more likely be CPU bound, if we
  have any performance issues with the Map Editor.

* Gamma correction seems to be done by the browser (so it has been removed)

* Convention for axis on texture and in WebGL are the same:
  - positive y-axis is vertical pointing upward
  - positive x-axis is horizontal pointing rightward

Nemikolh committed:
Add a way to preserve requestAnimationFrame

Zone.js is overidding this function, causing it to use
to much memory and be very slow. The production build
should use the native version so, this commit introduce
a hack to preserve the native version.

Nemikolh committed:
Fix naming: sprite files for TextureObject.

Nemikolh committed:
Start work on `Upload Chipset` action.

Nemikolh committed:
Fix a bug and set the default zoom to 2.

Nemikolh committed:
We can have handles on a layer and add a tile at a specific coordinate.

Nemikolh committed:
Fix issue with upload button.

Nemikolh committed:
Add VERSION and IS_PRODUCTION globals through webpack DefinePlugin.

Nemikolh committed:
Tiles can be updated by clicking on a particular location. Still buggy though.

Nemikolh committed:
Fix bugs with mouse handling.

Note:
  This commit introduce many hacks that needs
to be properly solved:
  - e.clientY is not the expected value.
  - tile_size is multiplied by 2 but should not.

Nemikolh commented on editor/rendering/tiles.ts:
```diff
@@ -0,0 +1,517 @@
+                for (let j = new_ij[1]; j < new_hw[1]; ++j) {
+                    let first_index = index / 2;
+
+                    vertices[index++] = this.pos[0] + 2 * this.tile_size * j - eps;
```
This `2` here is the hack mentioned in the previous commit message. The rendered size is incorrect without it. I haven't find the reason for the bug. It just doesn't make sense.

Nemikolh committed:
Zoom can now only have a set of interesting values.

Nemikolh committed:
Big clean up, Sprite oriented features!

 * SpriteRenderingContext and SpriteObject allow to
   manipulate single objects that use only one texture.
 * A blue-ish overlay can be added (done within the shader)
   used by the mouse.

Nemikolh committed:
Add basic support for the Palette.

This does not contains everything desired yet, but
introduces the main architecture to leave place for
the next features:

 * Select an area on the palette as a brush
 * Handle the editor state properly
   * Mouse/Key events
   * Actions
 * Layer selection

Nemikolh committed:
Add a cache to avoid using multiple texture ids for the same image.

Nemikolh committed:
Mouse position is now invariant when zooming in/out.

Nemikolh committed:
Maintaining the left click while moving, update the tiles values below the mouse.

Nemikolh committed:
Add pan behaviour + small refactoring.

Nemikolh committed:
Prepare the addition of the new Brush mechanism.

Nemikolh committed:
Small refactoring + support for texture of size 1.

Nemikolh committed:
Clean up + show the selected tile in Palette state.

Nemikolh committed:
Clean up, started work on LayersPanel.

Nemikolh committed:
Add support for multiple layers using the Map model.

Nemikolh committed:
Speed up the loop in dev.

Nemikolh committed:
We can now edit different layers! :)

Nemikolh committed:
Fixed bug when trying to pan on the Palette.

Nemikolh committed:
Hide editor content when the Palette is opened.

Nemikolh committed:
Fix brush bug.

Nemikolh committed:
Merge remote-tracking branch 'origin/master' into map-ed

Conflicts:
	editor/index.ts
	package.json

Nemikolh committed:
Fix compilation issues.

Nemikolh committed:
Fix angular2 issues with newest version.

ghost commented:
Looks like you are doing a great job here. I will try to test that as soon as possible!

Nemikolh commented:
Thanks a lot! :smiley_cat:  There's still plenty to do however.

I haven't documented all commands so here is a summary:
- Clicking on a valid location on the map replace the tile with the one under the mouse cursor.
  If the click is maintained and you move the mouse, it update all the tiles that are being encountered.
- If the map has the "focus" (can be required by clicking on the map at an invalid location), then pressing `P` will open the Palette (== virtual chipset described by @Azamung)
- You can zoom with the wheel and pan with the middle mouse button. There is two zoom behaviour: one for the map and one for the Palette.
- In the layers panel, selecting one of the fake image will change the current layer. All tile update will then be affected by this change.

ghost commented:
Okay. I plan to test all of that during before the end of the week.

Nemikolh committed:
Add createNewMap functionality.

Nemikolh committed:
Fix compilation bug (due to last angular2 upgrade.

ghost commented:
Okay so I tried to try, but I ran into an error. Both the server and the client spawned correctly, yet when I use my browser (url: `http://localhost:8001`), I got this error:

```
SyntaxError: in strict mode code, functions may be declared only at top level or immediately within another function index.js:24039:22
DEPRECATION WARNING: 'enqueueTask' is no longer supported and will be removed in next major release. Use addTask/addRepeatingTask/addMicroTask libs.js:58987:1
DEPRECATION WARNING: 'dequeueTask' is no longer supported and will be removed in next major release. Use removeTask/removeRepeatingTask/removeMicroTask libs.js:58987:1
[WDS] Warnings while compiling. libs.js:26356:4
./editor/rules/parser.ts
[26, 13]: unused variable: 'globalsInit'
[27, 13]: unused variable: 'locals'
 libs.js:26358:1
```

Note than when I did the `npm install`, I also got that:

```
npm WARN optional Skipping failed optional dependency /chokidar/fsevents:
npm WARN notsup Not compatible with your operating system or architecture: fsevents@1.0.8
npm WARN mongodb-core@1.2.31 requires a peer of kerberos@~0.0 but none was installed.                        
```

Probably the fault is mine, but if you could help me…

Nemikolh committed:
Fix a firefox bug.

With newer typescript, the generated code now includes 'use strict', changing the code semantics. :/

ghost commented:
Okay so your commit fixes the issue and now, it works (:

> Clicking on a valid location on the map replace the tile with the one under the mouse cursor. If the click is maintained and you move the mouse, it update all the tiles that are being encountered.

**Yes**, it works.

> If the map has the "focus" (can be required by clicking on the map at an invalid location), then pressing P will open the Palette

**Yes**.

> You can zoom with the wheel and pan with the middle mouse button. There is two zoom behaviour: one for the map and one for the Palette.

Sometimes, the edited map disappears and the only way to get it back is to zoom out. Zoom in after that makes the all thing disappars again.

> In the layers panel, selecting one of the fake image will change the current layer. All tile update will then be affected by this change.

![layer](https://cloud.githubusercontent.com/assets/13679561/13965984/cf0ec50a-f070-11e5-8c76-11e16b24efca.png)

I do not know if I am supposed to get that, but when I clicked on it, nothing _seems_ to happen... but maybe I missed something.

In any case, it is a pretty cool work you have done. Can’t wait to see your future work.

Nemikolh committed:
Remove unnecessary anonymous functions.

TypeScript generates them automatically now.

Nemikolh commented:
This PR fixes #9

Commit:
How to create an account with curl

close #10

Nemikolh committed:
Pixel perfect rendering!

* Refactor SceneManager into Pipeline.
* Show special screen when WebGL support is off.
* Add FBO to fix seams (to have seamless tilemap).
* Fix Socket.IO issue (#12).
* Partial addition of commit / new / get API for map.
* Fix README `lbkrb5-dev` missing.
* `gulpfile.js` fallback to dev mode by default.

Nemikolh commented:
Yay! Epic fight with my understanding of WebGL!! :smile: 

Last amend fix the annoying seams that were present in certain rendering conditions. This is something that is currently visible on the test instance:

![seams](https://cloud.githubusercontent.com/assets/4008156/15002876/e45205f0-119d-11e6-97e1-651ff24c78b9.png)

This is now fixed! We now have pixel perfect rendering for the map ! :tada: 

Well... almost.

There is a new issue with the texture rendering: the map orientation is wrong. However this should be relatively easier to fix and will come in another commit. :wink:

lthms commented:
You are awesome, man. Congrats!

Nemikolh committed:
Refactoring, Add a grid.

* 'Pipeline' is now more explicit
* 'Program' are now detached from '*RenderingContext'
* '*RenderingContext' have been renamed to '*RenderEl'
* Add a grid using a texture trick with GL_REPEAT.
  This impose zoom level to be powers of 2 only.
* Seam bug has been replaced by pixel errors now. This
  is quite annoying.

Nemikolh commented:
Okay, I'm not sure if this was a big win in definitive. Now you can randomly see this instead of the previous bug:

![new_bug](https://cloud.githubusercontent.com/assets/4008156/15052162/29eea5d0-12f4-11e6-8c8b-0a3c87de55f4.png)

![new_bug_2](https://cloud.githubusercontent.com/assets/4008156/15052230/7ac8f9f6-12f4-11e6-90aa-524a28c9bccd.png)

I will ask SO to see if someone can help me with that. I've tried every recommendations I've heard / read but nothing made a difference or fixed the bug (some produced even worst results).

lthms commented:
If I can help, just tell me.

Nemikolh commented:
Actually, if you could try on your machine to see if you can also see the bug (to trigger it you might need to zoom and pan a lot before it appears).

Thanks a lot!

lthms commented:
I will try as soon as I can, probably tomorrow night!

Nemikolh commented:
For reference, I posted the question here: http://gamedev.stackexchange.com/questions/121051/webgl-pixel-perfect-tilemap-rendering.

Let see if we can get any help! :smiley:

lthms commented:
Seems like it does not drow much crowds :\

Nemikolh committed:
Rewrite of TileLayer: Solve problems + perf gain!

 * `Pipeline` is renamed to `CommandBuffer`
 * Moved tile related stuff into its own folder
 * New rendering of tiles:
   * Draw only what's visible
   * No seams!
   * No pixels errors!
   * Super fast!!
   * =D
 * NOTE: at this point camera is totally broken. :'(

Nemikolh committed:
Fix previously introduced bugs with the new TileMap rendering.

Note:
  There is still a problem with the camera though. Will
  be fixed in a later commit.

lthms commented:
So you’ve found your way out? Congrats!

Nemikolh commented:
Yes! I found an article that contains a pure gpu solution. I originally thought that this would be a slow solution but I should have bench it first. :)
It is remarkably fast! :D

This is not yet finished though. I need to fix the camera and integrate the grid with the new technique. 

Finally I'll do some cleanup. The code has become very messy while I was experimenting.

Nemikolh committed:
Upgrade to angular2.0.0-beta.17.

Nemikolh committed:
Fix camera for new tile rendering.

Nemikolh committed:
Switch to port 8xxx for webpack-dev-server and server.

Nemikolh committed:
Libs updates. Now using latest version of Angular 2.

* This commit introduce an untested version
  of a webpack config for production.
* When the server is not there we see exception
  in the angular 2 http module.
* In general, errors are not properly handled.

Nemikolh committed:
Fix bugs related to lodash and typings upgrade.

lthms commented:
Do you think it’s a good idea to update the online version of dilia?

Nemikolh commented:
Not yet, I would like to update the sprite rendering first as it does not use the new "way" of computing position relatively to the camera. The two ways of rendering have a slight rendering offset resulting in shaky panning and incorrect overlay tiling.

lthms commented:
Sounds legit to me. Ping me when you want someone to test your awesome work (:

Nemikolh committed:
Improve by a little the build story.

Nemikolh committed:
Upgrade a few libs and fix compilation errors for the server.

Nemikolh committed:
Big angular update. Updated to latest router and also added AuthGuard to have proper redirection.

Nemikolh committed:
Fix the camera issue!

Nemikolh committed:
Fix the grid.

Nemikolh commented:
Thinking about it, this PR has been around for so long that it now changes so many part of the code base, it do so much more that was originally planned. Anything new that would done, actually need to be built on top of this PR. I'm thinking that we should merge it now and tackle each task listed here on their own, with small PR per task. That would be helpful to see progress more clearly.

@lethom : What do you think?

lthms commented:
You got me when you said _their own, with small PR per task_. Do you want me to do some tests here?

Nemikolh commented:
Yes please! That would be awesome! =)

If you don't find any "real" bug (only features missing), feel free to merge it.
If you find any though, I'll fix it here but without doing anything else (the cleanup will be my next PR).

Nemikolh committed:
Fix two tiny bugs: tiles fragment shader and type error.

Nemikolh committed:
Add the default chipset.

lthms commented:
Tonight is the night.
